File: camrod_system/scripts/diagnostics_aggregator_node.py
#!/usr/bin/env python3
# HH_260327: Diagnostics aggregator ported from todo/robot_diagnostics to camrod_system.

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

import rclpy
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus
from rclpy.node import Node


def _to_int_level(value: object) -> int:
    if isinstance(value, (bytes, bytearray)):
        if not value:
            return int(DiagnosticStatus.OK)
        return int(value[0])
    return int(value)


def _diag_level(value: int) -> bytes:
    return bytes([int(value) & 0xFF])


@dataclass
class Snapshot:
    status: DiagnosticStatus
    stamp_sec: float


class DiagnosticsAggregatorNode(Node):
    def __init__(self) -> None:
        super().__init__("diagnostics_aggregator")

        self.source_topic = str(
            self.declare_parameter("source_topic", "/diagnostics").value
        )
        self.output_topic = str(
            self.declare_parameter("output_topic", "/diagnostics_agg").value
        )
        self.publish_period_s = float(
            self.declare_parameter("publish_period_s", 1.0).value
        )
        self.stale_timeout_s = float(
            self.declare_parameter("stale_timeout_s", 3.0).value
        )

        self._snapshots: Dict[str, Snapshot] = {}

        self.sub = self.create_subscription(
            DiagnosticArray,
            self.source_topic,
            self._on_diag,
            20,
        )
        self.pub = self.create_publisher(DiagnosticArray, self.output_topic, 10)
        self.timer = self.create_timer(self.publish_period_s, self._on_timer)

        self.get_logger().info(
            "diagnostics_aggregator ready: "
            f"source={self.source_topic} output={self.output_topic} "
            f"stale_timeout_s={self.stale_timeout_s:.1f}"
        )

    def _copy_status(self, src: DiagnosticStatus) -> DiagnosticStatus:
        dst = DiagnosticStatus()
        dst.level = src.level
        dst.name = src.name
        dst.message = src.message
        dst.hardware_id = src.hardware_id
        dst.values = list(src.values)
        return dst

    def _on_diag(self, msg: DiagnosticArray) -> None:
        now_sec = self.get_clock().now().nanoseconds * 1e-9
        for status in msg.status:
            key = status.name if status.name else status.hardware_id
            if not key:
                continue
            self._snapshots[key] = Snapshot(
                status=self._copy_status(status),
                stamp_sec=now_sec,
            )

    def _on_timer(self) -> None:
        if not self._snapshots:
            return

        now = self.get_clock().now()
        now_sec = now.nanoseconds * 1e-9

        out = DiagnosticArray()
        out.header.stamp = now.to_msg()

        for key in sorted(self._snapshots.keys()):
            snap = self._snapshots[key]
            age = now_sec - snap.stamp_sec
            status = self._copy_status(snap.status)

            if age > self.stale_timeout_s:
                status.level = _diag_level(DiagnosticStatus.STALE)
                status.message = f"stale ({age:.1f}s)"

            out.status.append(status)

        self.pub.publish(out)


def main() -> None:
    rclpy.init()
    node = DiagnosticsAggregatorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: camrod_system/src/diagnostics_aggregator_node.cpp
#include <algorithm>
#include <cstdio>
#include <map>
#include <string>

#include "diagnostic_msgs/msg/diagnostic_array.hpp"
#include "diagnostic_msgs/msg/diagnostic_status.hpp"
#include "rclcpp/rclcpp.hpp"

namespace camrod_system
{

struct Snapshot
{
  diagnostic_msgs::msg::DiagnosticStatus status;
  double stamp_sec{0.0};
};

class DiagnosticsAggregatorNode : public rclcpp::Node
{
public:
  DiagnosticsAggregatorNode()
  : Node("diagnostics_aggregator")
  {
    source_topic_ = declare_parameter<std::string>("source_topic", "/diagnostics");
    // HH_260617: Default to a relative topic; system.launch.py namespaces it to
    // `/system/diagnostics_agg_tools` when used for lightweight system tools.
    output_topic_ = declare_parameter<std::string>("output_topic", "diagnostics_agg");
    publish_period_s_ = declare_parameter<double>("publish_period_s", 1.0);
    stale_timeout_s_ = declare_parameter<double>("stale_timeout_s", 3.0);

    sub_ = create_subscription<diagnostic_msgs::msg::DiagnosticArray>(
      source_topic_, 20,
      std::bind(&DiagnosticsAggregatorNode::on_diag, this, std::placeholders::_1));
    pub_ = create_publisher<diagnostic_msgs::msg::DiagnosticArray>(output_topic_, 10);
    timer_ = create_wall_timer(
      std::chrono::duration<double>(publish_period_s_),
      std::bind(&DiagnosticsAggregatorNode::on_timer, this));

    RCLCPP_INFO(
      get_logger(),
      "diagnostics_aggregator ready: source=%s output=%s stale_timeout_s=%.1f",
      source_topic_.c_str(), output_topic_.c_str(), stale_timeout_s_);
  }

private:
  // Copies only status payload fields used by downstream diagnostics consumers.
  static diagnostic_msgs::msg::DiagnosticStatus copy_status(
    const diagnostic_msgs::msg::DiagnosticStatus & src)
  {
    diagnostic_msgs::msg::DiagnosticStatus dst;
    dst.level = src.level;
    dst.name = src.name;
    dst.message = src.message;
    dst.hardware_id = src.hardware_id;
    dst.values = src.values;
    return dst;
  }

  // Stores latest status by stable key(name or hardware_id).
  void on_diag(const diagnostic_msgs::msg::DiagnosticArray::SharedPtr msg)
  {
    const double now_sec = now().seconds();
    for (const auto & status : msg->status) {
      const std::string key = !status.name.empty() ? status.name : status.hardware_id;
      if (key.empty()) {
        continue;
      }
      snapshots_[key] = Snapshot{copy_status(status), now_sec};
    }
  }

  // Publishes sorted aggregated status list with stale-state marking.
  void on_timer()
  {
    if (snapshots_.empty()) {
      return;
    }

    const auto stamp = now();
    const double now_sec = stamp.seconds();

    diagnostic_msgs::msg::DiagnosticArray out;
    out.header.stamp = stamp;

    for (const auto & kv : snapshots_) {
      const auto & snap = kv.second;
      const double age = now_sec - snap.stamp_sec;
      auto status = copy_status(snap.status);
      if (age > stale_timeout_s_) {
        status.level = diagnostic_msgs::msg::DiagnosticStatus::STALE;
        status.message = "stale (" + format_age(age) + "s)";
      }
      out.status.push_back(status);
    }

    pub_->publish(out);
  }

  // Formats age into fixed one-decimal text to keep legacy log style.
  static std::string format_age(double age_s)
  {
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%.1f", age_s);
    return std::string(buf);
  }

private:
  std::string source_topic_;
  std::string output_topic_;
  double publish_period_s_{1.0};
  double stale_timeout_s_{3.0};
  std::map<std::string, Snapshot> snapshots_;

  rclcpp::Subscription<diagnostic_msgs::msg::DiagnosticArray>::SharedPtr sub_;
  rclcpp::Publisher<diagnostic_msgs::msg::DiagnosticArray>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

}  // namespace camrod_system

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<camrod_system::DiagnosticsAggregatorNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
